### PR TITLE
Python: Add LOLWUT command

### PR DESCRIPTION
**Pull request by yipin-chen**
https://redis.io/docs/latest/commands/lolwut/

### CODE REVIEW

File: python/python/glide/async_commands/cluster_commands.py [hunk 1]
```diff
@@ -514,3 +514,42 @@ async def flushall(
...
+        parameters: Optional[List[int]] = None,
+        route: Optional[Route] = None,
+    ) -> str:
+        """
+        Displays a piece of generative computer art and the Redis version.
+
+        See https://valkey.io/commands/lolwut for more details.
+
+        Args:
+            version (Optional(int)): Version of computer art to generate.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/cluster_commands.py:530]: Same suggestion in other docs
```suggestion
            version (Optional[int]): Version of computer art to generate.
```

File: python/python/glide/async_commands/cluster_commands.py [hunk 2]
```diff
@@ -514,3 +514,42 @@ async def flushall(
...
+            parameters (Optional[str]): Additional set of arguments in order to change the output:
```
Comment (aaron-congo) [inline python/python/glide/async_commands/cluster_commands.py:531]: Same suggestion in other docs
```suggestion
            parameters (Optional[List[int]]): Additional set of arguments in order to change the output:
```

File: python/python/glide/async_commands/cluster_commands.py [hunk 3]
```diff
@@ -514,3 +514,42 @@ async def flushall(
...
+                For version <code>5</code>, those are length of the line, number of squares per row, and number of squares per column.
+                For version <code>6</code>, those are number of columns and number of lines.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/cluster_commands.py:533]: Same suggestion in other docs
```suggestion
                For version 5, those are length of the line, number of squares per row, and number of squares per column.
                For version 6, those are number of columns and number of lines.
```

File: python/python/glide/async_commands/cluster_commands.py [hunk 4]
```diff
@@ -514,3 +514,42 @@ async def flushall(
...
+            route (Optional[Route]): The command will be routed to all primary nodes, unless `route` is provided,
+                in which case the client will route the command to the nodes defined by `route`.
+
+        Returns:
+            str: A piece of generative computer art along with the current Redis version.
+
+        Examples:
+            >>>  client.lolwut(6, new int[] [ 40, 20 ], ALL_NODES);
```
Comment (aaron-congo) [inline python/python/glide/async_commands/cluster_commands.py:541]:
```suggestion
            >>>  client.lolwut(6, [40, 20], ALL_NODES);
```

File: python/python/glide/async_commands/cluster_commands.py [hunk 5]
```diff
@@ -514,3 +514,42 @@ async def flushall(
...
+            "Redis ver. 7.2.3" # Indicate the current Redis version
```
Comment (aaron-congo) [inline python/python/glide/async_commands/cluster_commands.py:542]: nit:
```suggestion
            "Redis ver. 7.2.3" # Indicates the current Redis version
```

File: python/python/glide/async_commands/cluster_commands.py [hunk 6]
```diff
@@ -514,3 +514,42 @@ async def flushall(
...
+
+        Since: Redis version 5.0.0.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/cluster_commands.py:544]: I think we only add this if the command was added in 6.0.0 or later, I don't see any notice about commands added in Redis 5 in any of the Java docs. Can be removed here and in the other docs

File: python/python/glide/async_commands/standalone_commands.py [hunk 7]
```diff
@@ -459,3 +459,42 @@ async def flushall(self, flush_mode: Optional[FlushMode] = None) -> TOK:
...
         )
+
+    async def lolwut(
+        self,
+        version: Optional[int] = None,
+        parameters: Optional[List[int]] = None,
+    ) -> str:
+        """
+        Displays a piece of generative computer art and the Redis version.
+        The command will be routed to a random node.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/standalone_commands.py:470]: this isn't applicable since its a standalone client
```suggestion
```

File: python/python/glide/async_commands/standalone_commands.py [hunk 8]
```diff
@@ -459,3 +459,42 @@ async def flushall(self, flush_mode: Optional[FlushMode] = None) -> TOK:
...
+                For version <code>6</code>, those are number of columns and number of lines.
+
+        Returns:
+            str: A piece of generative computer art along with the current Redis version.
+
+        Examples:
+            >>>  client.lolwut(6, new int[] [ 40, 20 ]);
+            "Redis ver. 7.2.3" # Indicate the current Redis version
+            >>> client.lolwut(5, new int[] [ 30, 5, 5 ]);
+            "Redis ver. 7.2.3" # Indicate the current Redis version
```
Comment (aaron-congo) [inline python/python/glide/async_commands/standalone_commands.py:487]:
```suggestion
            "Redis ver. 7.2.3" # Indicates the current Redis version
            >>> client.lolwut(5, new int[] [ 30, 5, 5 ]);
            "Redis ver. 7.2.3" # Indicates the current Redis version
```

File: python/python/glide/async_commands/transaction.py [hunk 9]
```diff
@@ -3442,6 +3442,34 @@ def getex(
...
         return self.append_command(RequestType.GetEx, args)
 
+    def lolwut(
+        self: TTransaction,
+        version: Optional[int] = None,
+        parameters: Optional[List[int]] = None,
+    ) -> TTransaction:
+        """
+        Displays a piece of generative computer art and the Redis version.
+        See https://valkey.io/commands/copy for more details.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/transaction.py:3452]: nit:
```suggestion
        
        See https://valkey.io/commands/copy for more details.
```

File: python/python/tests/test_async_client.py [hunk 10]
```diff
@@ -5559,6 +5559,49 @@ async def test_getex(self, redis_client: TRedisClient):
...
+                    assert "Redis ver. " in node_result
+            else:
+                assert "Redis ver. " in result
+
+            result = await redis_client.lolwut(2, [10, 20], RandomNode())
+            if isinstance(result, dict):
+                for node_result in result.values():
+                    assert "Redis ver. " in node_result
+            else:
+                assert "Redis ver. " in result
```
Comment (aaron-congo) [inline python/python/tests/test_async_client.py:5603]: I suggest altering this a bit since we know whether we should get back a dict or string based on the route we gave
```suggestion
            assert isinstance(result, dict)
            for node_result in result.values():
                assert "Redis ver. " in node_result

            result = await redis_client.lolwut(parameters=[10, 20], route=AllNodes())
            assert isinstance(result, dict):
            for node_result in result.values():
                assert "Redis ver. " in node_result

            # test with single-node route
            result = await redis_client.lolwut(2, route=RandomNode())
            assert "Redis ver. " in result

            result = await redis_client.lolwut(2, [10, 20], RandomNode())
            assert "Redis ver. " in result
```

File: python/python/tests/test_transaction.py [hunk 11]
```diff
@@ -763,3 +763,24 @@ async def test_transaction_lastsave(
...
+                if isinstance(node_result, str) and "Redis ver. " in node_result:
+                    assert True
+                else:
+                    assert False
+        else:
+            for element in results:
+                if isinstance(element, str) and "Redis ver. " in element:
+                    assert True
+                else:
+                    assert False
```
Comment (aaron-congo) [inline python/python/tests/test_transaction.py:786]: Similarly to the other tests, we know that we should not get a dict here because transaction should be sent to a single node, so we don't need the if-else
```suggestion
        for element in results:
            assert "Redis ver. " in element
```

File: python/python/glide/async_commands/cluster_commands.py [hunk 12]
```diff
@@ -514,3 +514,42 @@ async def flushall(
...
+        Displays a piece of generative computer art and the Redis version.
+
+        See https://valkey.io/commands/lolwut for more details.
+
+        Args:
+            version (Optional(int)): Version of computer art to generate.
+            parameters (Optional[str]): Additional set of arguments in order to change the output:
+                For version <code>5</code>, those are length of the line, number of squares per row, and number of squares per column.
+                For version <code>6</code>, those are number of columns and number of lines.
+            route (Optional[Route]): The command will be routed to all primary nodes, unless `route` is provided,
```
Comment (aaron-congo) [inline python/python/glide/async_commands/cluster_commands.py:534]: isn't it routed to a random node instead of all primaries?
```suggestion
            route (Optional[Route]): The command will be routed to a random node, unless `route` is provided,
```

File: python/python/tests/test_async_client.py [hunk 13]
```diff
@@ -5559,6 +5559,49 @@ async def test_getex(self, redis_client: TRedisClient):
...
+    async def test_lolwut(self, redis_client: TRedisClient):
+        result = await redis_client.lolwut()
+        assert "Redis ver. " in result
+        result = await redis_client.lolwut(parameters=[50, 20])
+        assert "Redis ver. " in result
+        result = await redis_client.lolwut(6)
+        assert "Redis ver. " in result
+        result = await redis_client.lolwut(5, [30, 4, 4])
+        assert "Redis ver. " in result
+
```
Comment (aaron-congo) [inline python/python/tests/test_async_client.py:5573]: Can we add a test for if an empty parameter list is passed?
```suggestion
        # passing an empty parameters list has no effect
        result = await redis_client.lolwut(parameters=[])
        assert "Redis ver. " in result
```

File: python/python/glide/async_commands/cluster_commands.py (hunk 4 above)
Comment (Yury-Fridlyand) [reply python/python/glide/async_commands/cluster_commands.py:541]: add `await`

File: python/python/glide/async_commands/cluster_commands.py (hunk 3 above)
Comment (Yury-Fridlyand) [reply python/python/glide/async_commands/cluster_commands.py:533]: Do we need to wrap numbers with `` ` ``?

File: python/python/glide/async_commands/standalone_commands.py [hunk 14]
```diff
@@ -459,3 +459,42 @@ async def flushall(self, flush_mode: Optional[FlushMode] = None) -> TOK:
...
+        Displays a piece of generative computer art and the Redis version.
+        The command will be routed to a random node.
+
+        See https://valkey.io/commands/lolwut for more details.
+
+        Args:
+            version (Optional(int)): Version of computer art to generate.
+            parameters (Optional[str]): Additional set of arguments in order to change the output:
+                For version <code>5</code>, those are length of the line, number of squares per row, and number of squares per column.
+                For version <code>6</code>, those are number of columns and number of lines.
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/standalone_commands.py:478]: same here

File: python/python/glide/async_commands/standalone_commands.py [hunk 15]
```diff
@@ -459,3 +459,42 @@ async def flushall(self, flush_mode: Optional[FlushMode] = None) -> TOK:
...
+
+        Returns:
+            str: A piece of generative computer art along with the current Redis version.
+
+        Examples:
+            >>>  client.lolwut(6, new int[] [ 40, 20 ]);
+            "Redis ver. 7.2.3" # Indicate the current Redis version
+            >>> client.lolwut(5, new int[] [ 30, 5, 5 ]);
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/standalone_commands.py:486]:
```suggestion
            >>> await client.lolwut(6, [ 40, 20 ]);
            "Redis ver. 7.2.3" # Indicate the current Redis version
            >>> await client.lolwut(5, [ 30, 5, 5 ]);
```

File: python/python/glide/async_commands/transaction.py [hunk 16]
```diff
@@ -3442,6 +3442,34 @@ def getex(
...
+
+        Args:
+            version (Optional(int)): Version of computer art to generate.
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/transaction.py:3455]:
```suggestion
            version (Optional[int]): Version of computer art to generate.
```

File: python/python/glide/async_commands/transaction.py [hunk 17]
```diff
@@ -3442,6 +3442,34 @@ def getex(
...
+            parameters (Optional[str]): Additional set of arguments in order to change the output:
+                For version <code>5</code>, those are length of the line, number of squares per row, and number of squares per column.
+                For version <code>6</code>, those are number of columns and number of lines.
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/transaction.py:3458]:
```suggestion
                For version 5, those are length of the line, number of squares per row, and number of squares per column.
                For version 6, those are number of columns and number of lines.
```
or
```suggestion
                For version `5`, those are length of the line, number of squares per row, and number of squares per column.
                For version `6`, those are number of columns and number of lines.
```

File: python/python/tests/test_async_client.py [hunk 18]
```diff
@@ -5559,6 +5559,49 @@ async def test_getex(self, redis_client: TRedisClient):
...
+        if isinstance(redis_client, RedisClusterClient):
```
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:5574]: Maybe move this piece of test into another test case which runs on cluster client only.

File: python/python/tests/test_transaction.py (hunk 11 above)
Comment (yipin-chen) [reply python/python/tests/test_transaction.py:786]: I may need to keep this `if isinstance(element, str)` otherwise CI/CD will fail

Comment (Yury-Fridlyand) [reply python/python/tests/test_transaction.py]: If it is not a `str`, so it should be a cluster response `Dict[str, str]` or so. So you should check all values in that `Dict`.

File: python/python/tests/test_async_client.py (hunk 18 above)
Comment (yipin-chen) [reply python/python/tests/test_async_client.py:5576]: Both standalone and cluster tests are not huge (after adopting Aaron's suggestions). I kind of want to keep them in the same test.

File: python/python/glide/async_commands/transaction.py [hunk 19]
```diff
@@ -3442,6 +3442,32 @@ def getex(
...
         return self.append_command(RequestType.GetEx, args)
 
+    def lolwut(
+        self: TTransaction,
+        version: Optional[int] = None,
+        parameters: Optional[List[int]] = None,
+    ) -> TTransaction:
+        """
+        Displays a piece of generative computer art and the Redis version.
+        See https://valkey.io/commands/lolwut for more details.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/transaction.py:3452]: nit: can you add an empty line above this?
```suggestion
        
        See https://valkey.io/commands/lolwut for more details.
```

File: python/python/tests/test_transaction.py (hunk 11 above)
Comment (aaron-congo) [reply python/python/tests/test_transaction.py]: This is a transaction, so it shouldn't be a cluster response I don't think

File: python/python/tests/test_transaction.py [hunk 20]
```diff
@@ -763,3 +763,15 @@ async def test_transaction_lastsave(
...
+    @pytest.mark.parametrize("cluster_mode", [True])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_lolwut_transaction(self, redis_client: RedisClusterClient):
+        transaction = Transaction()
+        transaction.lolwut().lolwut(5).lolwut(parameters=[1, 2]).lolwut(6, [42])
+        results = await redis_client.exec(transaction)
+        assert results is not None
+
+        for element in results:
+            if isinstance(element, str):
```
Comment (aaron-congo) [inline python/python/tests/test_transaction.py:776]: Do you get a mypy error if you remove this line? What about if you replace it with `assert isinstance(element, str)`?

Comment (yipin-chen) [reply python/python/tests/test_transaction.py:776]: If I remove this line, there is no issue on `mypy .`, but it will fail CI/CD. That is why I added it this morning.

**Review by aaron-congo — APPROVED** [review]

File: python/python/glide/async_commands/cluster_commands.py [hunk 21]
```diff
@@ -514,3 +514,40 @@ async def flushall(
             TClusterResponse[TOK],
             await self._execute_command(RequestType.FlushAll, args, route),
         )
+
+    async def lolwut(
+        self,
+        version: Optional[int] = None,
+        parameters: Optional[List[int]] = None,
+        route: Optional[Route] = None,
+    ) -> str:
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/cluster_commands.py:523]: In case of multi-node route, the client receives a map. Maybe this fix help you to eliminate `mypy` errors. Update tests accrodingly
```suggestion
    ) -> TClusterResponse[str]:
```

File: python/python/glide/async_commands/cluster_commands.py [hunk 22]
```diff
@@ -514,3 +514,40 @@ async def flushall(
...
+            "Redis ver. 7.2.3" # Indicates the current Redis version
+        """
+        args = []
+        if version is not None:
+            args.extend(["VERSION", str(version)])
+        if parameters:
+            for var in parameters:
+                args.extend(str(var))
+        return cast(
+            str,
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/cluster_commands.py:551]:
```suggestion
            TClusterResponse[str],
```

**Review by Yury-Fridlyand — APPROVED** [review]
Don't forget to update changelog on upstream
